docs/source/models/spatial-knn.rst

```diff
@@ -0,0 +1,288 @@
+* iteration: the iteration number
+* neighbour_number: the number of the neighbour in the K set
+* any other column from left dataset will be returned as well
+* any other column from right dataset will be returned as well
+* any column name that appears in both datasets will be suffixed with _right for the right dataset,
+    left dataset column names wont be altered
+
+For visualisation purposes we advise that you select the following columns:
+
+* left_id
+* right_id
+* geometry
+* right_geometry
+* geometry_geometry_distance
+
+The following image shows the result of the transformer applied on the buildings and taxi trip pickup locations:
+
+.. figure:: ../images/knn_result_visualisation.png
+   :figclass: doc-figure
+
+   Fig 3. Spatial KNN example visualisation.
+
+Mlflow Integration
+##################
+
+SpatialKNN transformer supports mlflow integration since it extends spark.mllib APIs.
+In addition the transformer comes with .getParams() and .getMetrics() methods to facilitate
+easy logging with mlflow. The .getParams() method returns a dictionary with the parameters
+of the transformer. The .getMetrics() method returns a dictionary with the metrics of the
+transformer after the convergence.
+
+.. tabs::
+   .. code-tab:: py
+
+    >>> import mosaic as mos
+    >>> mos.enable_mosaic(spark, dbutils)
+    >>>
+    >>> buildings_df = spark.read.table("buildings")
+    >>> buildings_df = buildings_df.withColumn("left_id", monotonically_increasing_id())
+    >>> roads_df = spark.read.table("roads")
+    >>> roads_df = roads_df.withColumn("right_id", monotonically_increasing_id())
+    >>>
+    >>> from mosaic.models import SpatialKnn
+    >>> import mlflow
+    >>>
+    >>> with mlflow.start_run():
+    >>>
+    >>>     knn = SpatialKNN(roads_df)
+    >>>     knn.setLeftGeometryColumn("geometry")
+    >>>     knn.setRightGeometryColumn("geometry")
+    >>>     knn.setLeftIdColumn("left_id")
+    >>>     knn.setRightIdColumn("right_id")
+    >>>     knn.setK(5)
+    >>>     knn.setMaxIterations(10)
+    >>>     knn.setDistanceThreshold(1.0)
+    >>>     knn.setCheckpointTablePrefix("checkpoint_table_knn")
+    >>>     knn.setIndexResolution(10)
+    >>>
+    >>>     neighbours = knn.transform(buildings_df)
+    >>>
+    >>>     mlflow.log_params(knn.getParams())
+    >>>     mlflow.log_metrics(knn.getMetrics())
+
+
+.. figure:: ../images/knn_mlflow_notebook.png
+   :figclass: doc-figure-full
+
+   Fig 4. Spatial KNN mlflow integration in notebooks.
+
+
+.. raw:: html
+
+   <div class="figure-group">
+
+
+.. figure:: ../images/knn_mlflow_params.png
+   :figclass: doc-figure-float-left
+
+   Fig 5. Spatial KNN mlflow integration params.
+
+.. figure:: ../images/knn_mlflow_metrics.png
+   :figclass: doc-figure-float-left
+
+   Fig 6. Spatial KNN mlflow integration metrics.
+
+
+.. raw:: html
+
+   </div>
+
+
+
+Shape Aware Hex Rings
+#####################
+
+When performing the iterations the transformer will use a hexagonal grid to identify
```

Comment:
Is it always hexagonal? What happens for rectangular grids? Is the k-ring in that case squared or does it approximate a circle?